src/test/java/org/junit/internal/runners/statements/FailOnTimeoutTest.java

```diff
@@ -34,176 +35,129 @@
-        assertEquals(TIMEOUT, e.getTimeout());
-        assertEquals(TimeUnit.MILLISECONDS, e.getTimeUnit());
-    }
-
-    private ThrowingRunnable evaluateWithDelegate(final Statement delegate) {
-        return new ThrowingRunnable() {
-            public void run() throws Throwable {
-                statement.nextStatement = delegate;
-                statement.waitDuration = 0;
-                failOnTimeout.evaluate();
-            }
-        };
-    }
-
-    private ThrowingRunnable evaluateWithException(Exception exception) {
-        return evaluateWithDelegate(new Fail(exception));
-    }
-
-    private ThrowingRunnable evaluateWithWaitDuration(final long waitDuration) {
-        return new ThrowingRunnable() {
-            public void run() throws Throwable {
-                statement.nextStatement = null;
-                statement.waitDuration = waitDuration;
-                failOnTimeout.evaluate();
-            }
-        };
-    }
-
-    private static final class TestStatement extends Statement {
-        long waitDuration;
-
-        Statement nextStatement;
-
-        @Override
-        public void evaluate() throws Throwable {
-            sleep(waitDuration);
-            if (nextStatement != null) {
-                nextStatement.evaluate();
-            }
-        }
+                run(failAfter50Ms(new InfiniteLoop())));
+        assertEquals(50, e.getTimeout());
+        assertEquals(MILLISECONDS, e.getTimeUnit());
     }
 
     @Test
     public void stopEndlessStatement() throws Throwable {
-        InfiniteLoopStatement infiniteLoop = new InfiniteLoopStatement();
-        FailOnTimeout infiniteLoopTimeout = builder().withTimeout(TIMEOUT, MILLISECONDS).build(infiniteLoop);
-        try {
-            infiniteLoopTimeout.evaluate();
-        } catch (Exception timeoutException) {
-            sleep(20); // time to interrupt the thread
-            int firstCount = InfiniteLoopStatement.COUNT;
-            sleep(20); // time to increment the count
-            assertTrue("Thread has not been stopped.",
-                    firstCount == InfiniteLoopStatement.COUNT);
-        }
-    }
-
-    private static final class InfiniteLoopStatement extends Statement {
-        private static int COUNT = 0;
-
-        @Override
-        public void evaluate() throws Throwable {
-            while (true) {
-                sleep(10); // sleep in order to enable interrupting thread
-                ++COUNT;
-            }
-        }
+        InfiniteLoop infiniteLoop = new InfiniteLoop();
+        assertThrows(
+                TestTimedOutException.class,
+                run(failAfter50Ms(infiniteLoop)));
+
+        sleep(20); // time to interrupt the thread
+        infiniteLoop.stillExecuting.set(false);
+        sleep(20); // time to increment the count
+        assertFalse(
+                "Thread has not been stopped.",
+                infiniteLoop.stillExecuting.get());
```

Comment:
I don't think this has a lot of value since Java does not allow stopping threads anyway. I'm more concerned about all tests that use this leaking threads. Can we tell `InfiniteLoop` instances to stop after we're done testing with them?

Reply:
I replaced InfiniteLoop with a statement that runs for 1 seconds so that it does not run forever. I also rename the test to make clear that it only stops statements that can be interrupted.